src/main/java/com/hivemq/persistence/local/memory/ClientQueueMemoryLocalPersistence.java

```diff
@@ -0,0 +1,988 @@
+                    final PublishWithRetained publish = (PublishWithRetained) messageWithID;
+                    retained = publish.retained;
+                    payloadPersistence.decrementReferenceCounter(Objects.requireNonNull(publish.getPayloadId()));
+                    increaseMessagesMemory(-publish.getEstimatedSize());
+                    pubrel.setExpiryInterval(publish.getMessageExpiryInterval());
+                    pubrel.setPublishTimestamp(publish.getTimestamp());
+                    replacedId = publish.getUniqueId();
+                } else if (messageWithID instanceof PubrelWithRetained) {
+                    final PubrelWithRetained pubrelWithRetained = (PubrelWithRetained) messageWithID;
+                    pubrel.setExpiryInterval(pubrelWithRetained.getExpiryInterval());
+                    pubrel.setPublishTimestamp(pubrelWithRetained.getPublishTimestamp());
+                    retained = pubrelWithRetained.retained;
+                }
+                break;
+            }
+        }
+        final PubrelWithRetained pubrelWithRetained = new PubrelWithRetained(pubrel, retained);
+        if(packetIdFound) {
+            messageQueue.set(messageIndexInQueue, pubrelWithRetained);
+        } else {
+            getOrPutQueueSize(key, bucketIndex).incrementAndGet();
+            // Ensure unknown PUBRELs are always first in queue
+            messageQueue.addFirst(pubrelWithRetained);
+        }
+        increaseMessagesMemory(pubrelWithRetained.getEstimatedSize());
+        return replacedId;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    @ExecuteInSingleWriter
+    public String remove(@NotNull final String client, final int packetId, final int bucketIndex) {
+        return remove(client, packetId, null, bucketIndex);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    @Nullable
+    @ExecuteInSingleWriter
+    public String remove(
+            @NotNull final String client, final int packetId, @Nullable final String uniqueId, final int bucketIndex) {
+        checkNotNull(client, "client id must not be null");
+        ThreadPreConditions.startsWith(SINGLE_WRITER_THREAD_PREFIX);
+
+        final Key key = new Key(client, false);
+        final Map<Key, LinkedList<MessageWithID>> bucket = buckets[bucketIndex];
+        final LinkedList<MessageWithID> messageQueue = getOrPutMessageQueue(key, bucket);
+        final Iterator<MessageWithID> iterator = messageQueue.iterator();
+        while(iterator.hasNext()){
+            final MessageWithID messageWithID = iterator.next();
+            if (messageWithID.getPacketIdentifier() == packetId) {
+                String removedId = null;
+                if (messageWithID instanceof PublishWithRetained) {
+                    final PublishWithRetained publish = (PublishWithRetained) messageWithID;
+                    if (uniqueId != null && !uniqueId.equals(publish.getUniqueId())) {
+                        break;
```

Comment:
is it really correct to break here, because it is not prevented in any way that another entry with another uniqueId could have the same packet-id